### ISSUE TITLE
"Could not initialize class io.netty.util.concurrent.DefaultPromise" when the slf4j implementation uses Netty

### ISSUE THREAD

**Issue by juherr** [netty/netty#5806]
I'm using a slf4j implementation which uses Netty and it fails at init with the following message:

```
Exception in thread "main" java.lang.NoClassDefFoundError: Could not initialize class io.netty.util.concurrent.DefaultPromise
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:35)
	at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:49)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:61)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:52)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:44)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:36)
```

As I understand the issue, at init Netty is trying to log into a logger which need Netty initialized! 🐔 
🍳 

I'm using Netty 4.0.x but the issue exists on 5.0.0 too.

You can reproduce it easily with:

```java
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Main {

    private static final Logger logger = LoggerFactory.getLogger(Main.class);

    public static void main(String... args) {
        logger.trace("BOOM");
    }
}
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>netty-issue</groupId>
    <artifactId>netty-issue</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.21</version>
        </dependency>
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-all</artifactId>
            <version>4.0.37.Final</version>
            <!--version>5.0.0.Alpha2</version-->
        </dependency>
    </dependencies>
</project>
```

```java
package org.slf4j.impl;

import io.netty.channel.nio.NioEventLoopGroup;
import org.slf4j.ILoggerFactory;
import org.slf4j.Logger;
import org.slf4j.Marker;
import org.slf4j.spi.LoggerFactoryBinder;

public final class StaticLoggerBinder implements LoggerFactoryBinder {

    private static final StaticLoggerBinder SINGLETON = new StaticLoggerBinder();

    public static StaticLoggerBinder getSingleton() {
        return SINGLETON;
    }

    @Override
    public ILoggerFactory getLoggerFactory() {
        NioEventLoopGroup group = new NioEventLoopGroup();
        return new ILoggerFactory() {
            @Override
            public Logger getLogger(String name) {
                return new Logger() {
                    @Override
                    public String getName() {
                        return name;
                    }

                    @Override
                    public boolean isTraceEnabled() {
                        return false;
                    }

                    @Override
                    public void trace(String msg) {
                    }

                    @Override
                    public void trace(String format, Object arg) {
                    }

                    @Override
                    public void trace(String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void trace(String format, Object... arguments) {
                    }

                    @Override
                    public void trace(String msg, Throwable t) {
                    }

                    @Override
                    public boolean isTraceEnabled(Marker marker) {
                        return true;
                    }

                    @Override
                    public void trace(Marker marker, String msg) {

                    }

                    @Override
                    public void trace(Marker marker, String format, Object arg) {
                    }

                    @Override
                    public void trace(Marker marker, String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void trace(Marker marker, String format, Object... argArray) {
                    }

                    @Override
                    public void trace(Marker marker, String msg, Throwable t) {
                    }

                    @Override
                    public boolean isDebugEnabled() {
                        return true;
                    }

                    @Override
                    public void debug(String msg) {
                    }

                    @Override
                    public void debug(String format, Object arg) {
                    }

                    @Override
                    public void debug(String format, Object arg1, Object arg2) {

                    }

                    @Override
                    public void debug(String format, Object... arguments) {
                    }

                    @Override
                    public void debug(String msg, Throwable t) {
                    }

                    @Override
                    public boolean isDebugEnabled(Marker marker) {
                        return true;
                    }

                    @Override
                    public void debug(Marker marker, String msg) {
                    }

                    @Override
                    public void debug(Marker marker, String format, Object arg) {
                    }

                    @Override
                    public void debug(Marker marker, String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void debug(Marker marker, String format, Object... arguments) {
                    }

                    @Override
                    public void debug(Marker marker, String msg, Throwable t) {
                    }

                    @Override
                    public boolean isInfoEnabled() {
                        return true;
                    }

                    @Override
                    public void info(String msg) {
                    }

                    @Override
                    public void info(String format, Object arg) {
                    }

                    @Override
                    public void info(String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void info(String format, Object... arguments) {
                    }

                    @Override
                    public void info(String msg, Throwable t) {
                    }

                    @Override
                    public boolean isInfoEnabled(Marker marker) {
                        return true;
                    }

                    @Override
                    public void info(Marker marker, String msg) {
                    }

                    @Override
                    public void info(Marker marker, String format, Object arg) {
                    }

                    @Override
                    public void info(Marker marker, String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void info(Marker marker, String format, Object... arguments) {
                    }

                    @Override
                    public void info(Marker marker, String msg, Throwable t) {
                    }

                    @Override
                    public boolean isWarnEnabled() {
                        return true;
                    }

                    @Override
                    public void warn(String msg) {
                    }

                    @Override
                    public void warn(String format, Object arg) {
                    }

                    @Override
                    public void warn(String format, Object... arguments) {
                    }

                    @Override
                    public void warn(String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void warn(String msg, Throwable t) {
                    }

                    @Override
                    public boolean isWarnEnabled(Marker marker) {
                        return true;
                    }

                    @Override
                    public void warn(Marker marker, String msg) {
                    }

                    @Override
                    public void warn(Marker marker, String format, Object arg) {
                    }

                    @Override
                    public void warn(Marker marker, String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void warn(Marker marker, String format, Object... arguments) {
                    }

                    @Override
                    public void warn(Marker marker, String msg, Throwable t) {
                    }

                    @Override
                    public boolean isErrorEnabled() {
                        return true;
                    }

                    @Override
                    public void error(String msg) {
                    }

                    @Override
                    public void error(String format, Object arg) {
                    }

                    @Override
                    public void error(String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void error(String format, Object... arguments) {
                    }

                    @Override
                    public void error(String msg, Throwable t) {
                    }

                    @Override
                    public boolean isErrorEnabled(Marker marker) {
                        return true;
                    }

                    @Override
                    public void error(Marker marker, String msg) {
                    }

                    @Override
                    public void error(Marker marker, String format, Object arg) {
                    }

                    @Override
                    public void error(Marker marker, String format, Object arg1, Object arg2) {
                    }

                    @Override
                    public void error(Marker marker, String format, Object... arguments) {
                    }

                    @Override
                    public void error(Marker marker, String msg, Throwable t) {
                    }
                };
            }
        };
    }

    @Override
    public String getLoggerFactoryClassStr() {
        return "";
    }
}
```

**Comment by Scottmitch** [comment]
Yes there appears to be a cycle in the class loading. Can you explain why you have `NioEventLoopGroup group = new NioEventLoopGroup();` while you are initializing your logger?

**Comment by juherr** [comment]
It comes from io.netty.util.concurrent.DefaultPromise hierarchie. 
I found it when I tried to understand the root cause of the exception.

**Comment by Scottmitch** [comment]
What I'm wondering is why you have the following line in `getLoggerFactory()`:

```java
NioEventLoopGroup group = new NioEventLoopGroup();
```

It doesn't appear to be used and it is creating a cycle during class loading. If I remove this things work as expected.

**Comment by juherr** [comment]
I don't know. The logger I use is https://github.com/Graylog2/log4j2-gelf/issues/22 .

**Comment by juherr** [comment]
Well, I checked the issue again and in my real life scenario the one which is creating `NioEventLoopGroup` is Apache CXF:

```
Exception in thread "main" java.lang.NoClassDefFoundError: Could not initialize class io.netty.channel.nio.NioEventLoopGroup
	at org.apache.cxf.transport.http.netty.client.NettyHttpConduitFactory.createConduit(NettyHttpConduitFactory.java:89)
	at org.apache.cxf.transport.http.HTTPTransportFactory.getConduit(HTTPTransportFactory.java:226)
	at org.apache.cxf.transport.http.HTTPTransportFactory.getConduit(HTTPTransportFactory.java:208)
	at org.apache.cxf.transport.TransportURIResolver.resolve(TransportURIResolver.java:109)
	at org.apache.cxf.wsdl11.CatalogWSDLLocator.getBaseInputSource(CatalogWSDLLocator.java:74)
	at org.apache.cxf.wsdl11.AbstractWrapperWSDLLocator.getBaseInputSource(AbstractWrapperWSDLLocator.java:57)
	at org.apache.cxf.wsdl11.WSDLManagerImpl.loadDefinition(WSDLManagerImpl.java:209)
	at org.apache.cxf.wsdl11.WSDLManagerImpl.getDefinition(WSDLManagerImpl.java:163)
	at org.apache.cxf.wsdl11.WSDLServiceFactory.<init>(WSDLServiceFactory.java:85)
	at org.apache.cxf.jaxws.ServiceImpl.initializePorts(ServiceImpl.java:218)
	at org.apache.cxf.jaxws.ServiceImpl.initialize(ServiceImpl.java:161)
	at org.apache.cxf.jaxws.ServiceImpl.<init>(ServiceImpl.java:129)
	at org.apache.cxf.jaxws.spi.ProviderImpl.createServiceDelegate(ProviderImpl.java:82)
```

I'm using CXF 3.1.7.

I hope it will help.

**Comment by Scottmitch** [comment]
Have you reached out to the Apache CXF folks (I'm not familiar with this code base)?

**Comment by juherr** [comment]
Not yet, but I'm not able to explain to them what they are doing wrong. They are just creating a `NioEventLoopGroup` and it is failing (what you can reproduce with my sample) in a very specific case: cxf -> netty -> slf4j -> slf4j-to-log2j2 -> log4j2-gelf -> netty (or more generic: cxf -> netty -> slf4j impl using netty -> netty).

If they are not supposed to use `NioEventLoopGroup` for any reason, I think you're a better person to explain to them why ;)

**Comment by Scottmitch** [comment]
In the example code you provided above there is cyclic dependency bcz you are creating a `NioEventLoopGroup` while you are statically loading your logger, but as you indicated `NioEventLoopGroup` depends upon the logger at class load time (cycle). I'm not sure if `CXF` is doing the same thing ... but if so I would also expect a cyclic dependency.

Using [InternalLoggerFactory#setDefaultFactory(..)](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java#L79) to initialize/override the internal logger maybe easier to reason about.

**Comment by juherr** [comment]
You can have a look on CXF usage here: https://github.com/apache/cxf/blob/ac9b9b1898bbe0cc911b34cc3d32664ac59fdc34/rt/transports/http-netty/netty-client/src/main/java/org/apache/cxf/transport/http/netty/client/NettyHttpConduitFactory.java#L89

**Comment by juherr** [comment]
But the problem occurs when `NioEventLoopGroup` is called before logger initialization.
See my sample on https://github.com/Graylog2/log4j2-gelf/issues/22
In my true life use case, CXF is just replacing the main.

Fyi, with log4j2-gelf logger impl, something like 

```java
        NioEventLoopGroup group = new NioEventLoopGroup();
        Logger logger = LogManager.getLogger("test");
```

will fail, with:

```
java.lang.NoClassDefFoundError: Could not initialize class io.netty.util.concurrent.DefaultPromise

	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:35)
	at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:49)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:61)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:52)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:44)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:36)
[...]
```

when

```java
        Logger logger = LogManager.getLogger("test");
        NioEventLoopGroup group = new NioEventLoopGroup();
```

is working as expected

```
2016-09-15 21:51:43,424 main ERROR Recursive call to appender gelfAppender
2016-09-15 21:51:43.424 [DEBUG] [main] [o.g.g.t.AbstractGelfTransport] - Trying to send message: GelfMessage{version="1.1" timestamp="1473969103,422" short_message="Instrumented an optimized java.util.Set into: sun.nio.ch.EPollSelectorImpl@67e2d983", level="DEBUG(7)"}
2016-09-15 21:51:43.422 [TRACE] [main] [i.n.c.n.NioEventLoop] - Instrumented an optimized java.util.Set into: sun.nio.ch.EPollSelectorImpl@67e2d983
[...]
```

As workarroud, I may try to init logger on my side before everything else (I will it try tomorow)

**Comment by juherr** [comment]
The workaround is working.

**Comment by Scottmitch** [comment]
good to hear you are making progress ... I'm not convinced this is a Netty issue yet. If you can provide a reproducer with just Netty that would help me understand the issue more.

**Comment by juherr** [comment]
I've already provided the sample :)

The problem only occurs when a logger implementation and the business code use netty AND only if the logger is initialized after the first usage of netty.

I don't know if we can call it a "netty issue" but I think netty is the best place to make a fix.

**Comment by Scottmitch** [comment]
> I've already provided the sample :)

I've already explained what is wrong with the example in https://github.com/netty/netty/issues/5806#issuecomment-247400442.

**Comment by juherr** [comment]
Just apply the provided patch on the `4.1` branch and try to run `EchoServer`.
[Netty_issue__5806.txt](https://github.com/netty/netty/files/491122/Netty_issue__5806.txt)

Is it clear enough for you now?

**Comment by Scottmitch** [comment]
Yes this is helpful. The initialization is a bit incestuousness in that the statically initialized logger is attempting to use Netty, and Netty statically needs to use the logger which is being initialized. However I think I can come up with a PR to prevent the exception ... standby

**Comment by Scottmitch** [comment]
I was able to push off the static exception for a bit, but I don't think the way `gelfclient` is being used will work because as previously stated there is a cycle during static initialization. After I was able to push off the first issue more issues related to static state used by netty will pop up. For example [AbstractGelfTransport](https://github.com/Graylog2/gelfclient/blob/master/src/main/java/org/graylog2/gelfclient/transport/AbstractGelfTransport.java#L41) attempts to use the "default number of threads" by passing in `0`, but translating the `0` into the default value depends upon a static code block to run in [MultithreadEventLoopGroup](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/MultithreadEventLoopGroup.java#L51) ... but it hasn't yet run because the logger is trying to be statically initialized at this time (cycle see exception below).

```
java.lang.IllegalArgumentException: nThreads: 0 (expected: > 0)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:72)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:58)
	at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:47)
	at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:58)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:77)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:72)
	at io.netty.channel.nio.NioEventLoopGroup.<init>(NioEventLoopGroup.java:59)
	at org.graylog2.gelfclient.transport.AbstractGelfTransport.<init>(AbstractGelfTransport.java:41)
	at org.graylog2.gelfclient.transport.AbstractGelfTransport.<init>(AbstractGelfTransport.java:62)
	at org.graylog2.gelfclient.transport.GelfUdpTransport.<init>(GelfUdpTransport.java:49)
	at org.graylog2.gelfclient.GelfTransports.create(GelfTransports.java:47)
	at org.graylog2.gelfclient.GelfTransports.create(GelfTransports.java:68)
	at org.graylog2.log4j2.GelfAppender.start(GelfAppender.java:186)
	at org.apache.logging.log4j.core.config.AbstractConfiguration.start(AbstractConfiguration.java:247)
	at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:496)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:566)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:582)
	at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:217)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:152)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:45)
	at org.apache.logging.log4j.LogManager.getContext(LogManager.java:194)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getContext(AbstractLoggerAdapter.java:103)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getContext(Log4jLoggerFactory.java:43)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getLogger(AbstractLoggerAdapter.java:42)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getLogger(Log4jLoggerFactory.java:29)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:358)
	at io.netty.util.internal.logging.Slf4JLoggerFactory.newInstance(Slf4JLoggerFactory.java:73)
	at io.netty.util.internal.logging.InternalLoggerFactory.newDefaultFactory(InternalLoggerFactory.java:54)
	at io.netty.util.internal.logging.InternalLoggerFactory.getDefaultFactory(InternalLoggerFactory.java:77)
	at io.netty.util.internal.logging.InternalLoggerFactory.getInstance(InternalLoggerFactory.java:103)
	at io.netty.util.internal.logging.InternalLoggerFactory.getInstance(InternalLoggerFactory.java:96)
	at io.netty.util.internal.ThreadLocalRandom.<clinit>(ThreadLocalRandom.java:65)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at io.netty.util.internal.logging.InternalLoggerFactory.<clinit>(InternalLoggerFactory.java:43)
	at io.netty.channel.MultithreadEventLoopGroup.<clinit>(MultithreadEventLoopGroup.java:34)
	at io.netty.example.echo.EchoServer.main(EchoServer.java:52)
```

gelfclient also depends upon Netty 4.0.37. I wouldn't recommend attempting to have a 4.0.37 dependency resolve to a 4.1 Netty. It may make more sense for gelfclient to bring its own Netty (via shading etc...) and not attempt to share with the current application. Plz discuss with the maintainer of gelfclient.

```
[INFO] +- org.graylog2.log4j2:log4j2-gelf:jar:1.3.1-SNAPSHOT:runtime
[INFO] |  \- org.graylog2:gelfclient:jar:1.4.1-SNAPSHOT:runtime
[INFO] |     +- (io.netty:netty-example:jar:4.0.37.Final:runtime - omitted for cycle)
[INFO] |     +- io.netty:netty-all:jar:4.0.37.Final:runtime
[INFO] |     +- com.fasterxml.jackson.core:jackson-core:jar:2.5.4:runtime
[INFO] |     \- (org.slf4j:slf4j-api:jar:1.7.21:runtime - version managed from 1.7.12; omitted for duplicate)
```

**Comment by juherr** [comment]
In fact, they already have a shaded artifact: http://search.maven.org/#search%7Cga%7C1%7Clog4j2-gelf

And it looks working as expected with:

```xml
    <dependency>
      <groupId>org.graylog2.log4j2</groupId>
      <artifactId>log4j2-gelf</artifactId>
      <version>1.3.0</version>
      <classifier>shaded</classifier>
      <scope>runtime</scope>
    </dependency>
```

**Comment by Scottmitch** [comment]
to clarify using the shaded jar resolved this issue?

**Comment by Scottmitch** [comment]
@juherr - I will close this issue for now. Plz reopen if you feel there is something more we can do in Netty.

**Comment by rgolangh** [comment]
@Scottmitch I am seeing the same error
`java.lang.NoClassDefFoundError: Could not initialize class io.netty.util.concurrent.DefaultPromise`

But the underlying issue is within https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/PlatformDependent0.java#L107 - if for some reason you can't load Unsafe, ClassNotFoundException will be thrown, and that will propagate all the way up and fail DefaultPromise class init. 

There should be a fix to catch CNFE and set the `maybeUnsafe` with it and that will be consistent with the rest of the of the treatment. But I'm not sure what is the behaviour with missing unsafe. At least I would like to know specifically that it can't be reached. If this is the direction let me know and I'll send the fix.

Meanwhile, since I'm using wildfly, I had to add to my module a dependency upon `sun.jdk` to solve it , thanks to [this answer](http://stackoverflow.com/a/31589915)
```
<module name="sun.jdk" export="true">
  <imports>
    <include path="sun/misc/Unsafe"/>
  </imports>
</module>
```

**Comment by jasontedor** [comment]
If you can never load unsafe, you should just disable it: `-Dio.netty.noUnsafe=true` or otherwise add the necessary configuration as you've done.

**Comment by Scottmitch** [comment]
@rgolangh - What is your JDK version and OS details?

Yes it seems like we should catch CNFE in this case (and potentially others where we do similar reflection based operations). @jasontedor - any objection to catching `Exception` here ... or just keep adding individual exceptions as we discover them?

**Comment by jasontedor** [comment]
@Scottmitch Rather than adding exceptions, I'd go the other direction. Consider how this could have gone down had `ClassNotFoundException` been handled the way that the required checked exceptions are handled. The developer here would not know that unsafe was unavailable unless they happened to have enabled debug logging and were reading debug log messages. Because of the non-silent failure, the developer was made aware of the situation and able to rectify it accordingly. My position would be that instead of just logging the exceptions that are currently logged, just throw them (as unchecked exceptions) to remove any silent failure here (I consider logging silent failure, no one reads logs unless absolutely necessary) and I would definitely not add more exceptions to silently fail on.

**Comment by Scottmitch** [comment]
`Unsafe` is optional and does not have to be considered a fatal failure on startup (in general it is intended to be an optimization). We take measures to ensure the startup sequence can continue on if `Unsafe` is not present and CNFE happens to be a hole. If the user wants to make `Unsafe` mandatory and fail the start up sequence the user is free to impose this restriction at a higher level.

For additional context `Unsafe` is required for the JNI transports (epoll / kqueue) and will result in a failure if these transports are used and `Unsafe` is not available.

**Comment by jasontedor** [comment]
I know it's optional, but that doesn't negate anything that I said about silent failures.

**Comment by Scottmitch** [comment]
> I know it's optional, but that doesn't negate anything that I said about silent failures.

Exception handling seems to be a reoccurring theme in our discussions :) In general I agree that "silent failures", "swallowing exceptions", and otherwise reducing visibility for issues that the user must deal with is a not a good idea. In this specific case the user does not necessarily have to know/care that we use `Unsafe` under the hood. Using `Unsafe` is an internal implementation detail of Netty and not required to use Netty. Given the optional nature of `Unsafe` it seems excessive to prevent the application (which may do other stuff besides just Netty) from starting. WDYT?

**Comment by rgolangh** [comment]
@Scottmitch I'm using openjdk 1.8.0_111-b16 on Fedora 25

Can we couple the check of Unsafe with epoll/kqueue usage? if that's clean enough to do then I think its the best of both worlds, explode netty if you have a core dependency missing or just a WARN msg in case its not there.

**Comment by jasontedor** [comment]
> Exception handling seems to be a reoccurring theme in our discussions :)

@Scottmitch Yes. I think silent or late failures are a scourge that just lead to harder-to-debug problems down the line.

> WDYT?

I hear you, but I'm unswayed. In addition to the places you list where unsafe is required, it is also used as an optional performance optimizations in various places. Netty has presumably deemed these optimizations important enough that unsafe is enabled by default. Not explicitly failing here just makes, for example, performance issues, much harder to track down.

Either:

1. these performance optimizations are important enough to enable unsafe by default in which case Netty should explicitly fail if unsafe is unavailable unless disabled
2. these performance optimizations are not important enough to enable unsafe by default, in which case it should be disabled by default and then if enabled by the user of course Netty should fail if unavailable

In short, if it's supposed to be on (by default, or explicitly), fail explicitly if it's not available.

> Given the optional nature of `Unsafe` it seems excessive to prevent the application (which may do other stuff besides just Netty) from starting.

There's a simple out though: disable unsafe explicitly.

**Comment by normanmaurer** [comment]
@jasontedor @Scottmitch sorry for joining late but my vacation was too good ;) 

@jasontedor I disagree with you to some extend here. I think if netty can be used (and in fact it does) without Unsafe it is good enough to just log and move on. I think require the user to explicit set a system property on some systems and on some systems not is worse then "automatically" handle it.

**Comment by Scottmitch** [comment]
> Can we couple the check of Unsafe with epoll/kqueue usage?

@rgolangh - Unsafe should already be required to use kqueue and epoll native transports. If we are not enforcing this correctly feel free to file an issue.

**Comment by Scottmitch** [comment]
@jasontedor - Netty strives to provide "good performance" in "most use cases". Netty is widely used for many different protocols, and in many different use cases. Generally when using `Unsafe` is a reasonable option it can provide a performance boost (unless intrinsics come into play), but we can't speak for every JDK and every platform that folks may use. If we suspect (or are told) this is not the case we will investigate and consider changing the default for that specific case.  Applications/libraries on top of Netty are free to be explicit about defaults and take action like fail application start up if improperly configured where it assumed more is known about environment and performance behaviors.
